File: solutions/141.linked-list-cycle/linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
function hasCycle(head) {
  if(!head) return false
    
  let p1 = head
  let p2 = head

  while (p2.next !== null && p2.next.next !== null) {
    p1 = p1.next
    p2 = p2.next.next
    if (p1.val === p2.val) return true
  }

  return false
}

File: solutions/141.linked-list-cycle/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head:
            return False

        slow = head
        fast = head.next

        while fast and fast.next:
            if slow.val == fast.val:
                return True

            slow = slow.next
            fast = fast.next.next

        return False
